### PR TITLE
Port ec2_tag to boto3

### PR DESCRIPTION
**Pull request by flowerysong**
##### SUMMARY

Ports ec2_tag to boto3 and AnsibleAWSModule.

Fixes authentication failures when using temporary credentials from an assumed role.

Also adds `purge_tags`, in line with other AWS modules that manipulate tags.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

ec2_tag

##### ANSIBLE VERSION
2.7

##### ADDITIONAL INFORMATION

Example boto auth failure:
```
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_TPlP85/ansible_module_ec2_tag.py\", line 183, in <module>\n    main()\n  File \"/tmp/ansible_TPlP85/ansible_module_ec2_tag.py\", line 141, in main\n    gettags = ec2.get_all_tags(filters=filters)\n  File \"/home/zeke/git/ansible-aws/bin/python/local/lib/python2.7/site-packages/boto/ec2/connection.py\", line 4195, in get_all_tags\n    [('item', Tag)], verb='POST')\n  File \"/home/zeke/git/ansible-aws/bin/python/local/lib/python2.7/site-packages/boto/connection.py\", line 1186, in get_list\n    raise self.ResponseError(response.status, response.reason, body)\nboto.exception.EC2ResponseError: EC2ResponseError: 401 Unauthorized\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors><Error><Code>AuthFailure</Code><Message>AWS was not able to validate the provided access credentials</Message></Error></Errors><RequestID>e6dcd418-1814-41d9-a379-c8b63e23a309</RequestID></Response>\n",
```

### CODE REVIEW

**Comment by flowerysong** [comment]
Probably needs #39765 for the integration tests to work in CI.

**Comment by mattclay** [comment]
The CI failure appears to be due to lack of permissions in CI.

cc @s-hertel

File: lib/ansible/modules/cloud/amazon/ec2_tag.py [hunk 1]
```diff
@@ -33,8 +34,17 @@
     description:
       - a hash/dictionary of tags to add to the resource; '{"key":"value"}' and '{"key":"value","key":"value"}'
     required: true
-
-author: "Lester Wade (@lwade)"
+  purge_tags:
+    description:
+      - Whether unspecified tags should be removed from the resource.
+      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_tag.py:40]: I think this behavior should be shown in the examples since it is tricky to explain.

**Review by ryansb — DISMISSED** [review]
Minor docs request, this refactor looks good overall. Thanks for the added tests as well!

File: lib/ansible/modules/cloud/amazon/ec2_tag.py (hunk 1 above)
Comment (flowerysong) [reply lib/ansible/modules/cloud/amazon/ec2_tag.py:40]: Do we need an additional example beyond lines 80-87? (Or is there anything about it you would change?)
```
+- name: Remove all tags except for Name from an instance
+  ec2_tag:
+    region: eu-west-1
+    resource: i-xxxxxxxxxxxxxxxxx
+    tags:
+        Name: ''
+    state: absent
+    purge_tags: true
```

File: lib/ansible/modules/cloud/amazon/ec2_tag.py [hunk 2]
```diff
@@ -90,93 +70,115 @@
...
+    resource = module.params['resource']
+    tags = module.params['tags']
+    state = module.params['state']
+    purge_tags = module.params['purge_tags']
+
+    result = {'changed': False}
+
+    try:
+        ec2 = module.client('ec2')
+    except (BotoCoreError, ClientError) as e:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_tag.py]: The connection doesn't need exception handling.

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
bot_status

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
rebuild_merge

**Pull request merged**